### ISSUE TITLE
Including text metadata in pinecone upsert.

### ISSUE THREAD

**Issue by SnoozingSimian** [langchain-ai/langchain#1341]
While using langchain for pinecone upsert I was frequently running into an error discussed [here](https://community.pinecone.io/t/metadata-size-error/604/11). I found that langchain is including text metadata along with whatever the user sends as metadata [here](https://github.com/hwchase17/langchain/blob/master/langchain/vectorstores/pinecone.py#:~:text=for%20j%2C,%5D%20%3D%20line). What is the purpose of doing this? It will easily generate metadatas which are over the permissible [limit set by pinecone](https://docs.pinecone.io/docs/limits#metadata:~:text=Max%20metadata%20size%20per%20vector%20is%2010%20KB.) in many cases.

**Comment by SnoozingSimian** [comment]
Even though this might still cause high cardinality for metadata, I guess this step is essential in order to actually fetch the text.

**Comment by kitfit-dave** [comment]
I have the same question here too. What are the thoughts on doing this in a way that will not cause high cardinality? I am basically thinking that pinecone (or any vector database) should be storing the vectors and an id, and that the texts should be elsewhere indexes by that id. I do think that other vector databases (e.g. weaviate) deal with having the text alongside the vector data better than pinecone claims to - perhaps the idea is just to wait it out and eventually storing the text will be okay?

**Comment by kitfit-dave** [comment]
I do note that Pinecone, in their examples (e.g. https://docs.pinecone.io/docs/semantic-text-search#creating-an-index or https://docs.pinecone.io/docs/gen-qa-openai#indexing-data-in-vector-db etc.) do just go ahead and store the text in the metadata, rather than "storing high cardinality data in a separate key value store" as they suggest elsewhere (https://docs.pinecone.io/docs/troubleshooting#pods-are-full)

**Comment by kitfit-dave** [comment]
TL;DR: make sure you create your index specifying only the metadata fields other than "text" (where the metadata will be put by default) for indexing

I also had the same question. I did find that pinecone have many examples of putting the text in the metadata (e.g. https://docs.pinecone.io/docs/semantic-text-search#creating-an-index or https://docs.pinecone.io/docs/gen-qa-openai#indexing-data-in-vector-db etc.). They do suggest putting high cardinality data in the metadata for filtering is not a good idea (https://docs.pinecone.io/docs/troubleshooting#pods-are-full), they do have a solution for storing metadata that is not filterable: Selective metadata indexing (https://docs.pinecone.io/docs/manage-indexes#selective-metadata-indexing), which should be fine for us as the langchain code does not ever create the index, so one should be able to specify that the text metadata column should not be indexed when creating the index.

I haven't seen this pointed out in any pinecone docs, though it surely must cause issues for anyone just working from their examples. It might also be a good idea to note this somewhere in the langchain documentation specific to pinecone, that the index should be set up to not index the metadata field where the text will be stored.

This might become more of a concern if langchain ever decides to create indexes, or more deeply integrate searching the metadata - would then need to ensure metadata_config was set up correctly.

(I've combined my two previous comments, posted without fully investigating, into this summary one)

**Comment by amuhareb** [comment]
@kitfit-dave thanks so much for this explanation. My index of embeddings was created elsewhere and I'm not keen on rebuilding it, so i think I understand now how I can modify my index to have metadata. It may not be recommended, but it'll suit my purpose especially as I'm just learning and testing for now.

**Comment by amuhareb** [comment]
@kitfit-dave thanks so much for this explanation. My index of embeddings was created elsewhere and I'm not keen on rebuilding it, so i think I understand now how I can modify my index to have metadata. It may not be recommended, but it'll suit my purpose especially as I'm just learning and testing for now.
